src/main/utility/mod.rs

```diff
@@ -44,3 +44,38 @@ pub trait IsSend: Send {}
 
 /// A trait we can use as a compile-time check to make sure that an object is Sync.
 pub trait IsSync: Sync {}
+
+/// Runtime memory error checking to help catch errors that C code is prone
+/// to. Can probably drop once C interop is removed.
+///
+/// The MAGIC parameter should ideally be unique for each type. Consider e.g.
+/// `python3 -c 'import random; print(random.randint(0, 2**32))'`
+#[derive(Debug, Clone)]
+pub struct Magic<const MAGIC: u32> {
+    #[cfg(debug_assertions)]
+    magic: u32,
+}
+
+impl<const MAGIC: u32> Magic<MAGIC> {
+    pub fn new() -> Self {
+        Self {
+            #[cfg(debug_assertions)]
+            magic: MAGIC,
+        }
+    }
+
+    pub fn debug_check(&self) {
+        #[cfg(debug_assertions)]
+        debug_assert_eq!(unsafe { std::ptr::read_volatile(&self.magic) }, MAGIC);
+    }
```

Comment:
I think we also want a `std::sync::atomic::compiler_fence` here to make sure that the compiler doesn't reorder reads. For example, we probably want to make sure that the magic value is read before any other struct field is read.

Reply:
Good point; done.